// react-cloudinary-lazy-image/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var makeUrlParams = function makeUrlParams(props) {
  var urlParams = props.urlParams,
      imgFormat = props.imgFormat,
      quality = props.quality,
      fluid = props.fluid,
      blurUrlParams = props.blurUrlParams,
      useUrlParamsToBlur = props.useUrlParamsToBlur,
      blurSize = props.blurSize;
  imgFormat = typeof imgFormat === "boolean" ? imgFormat ? "f_auto" : '' : imgFormat;
  quality = typeof quality === "boolean" ? quality ? "q_auto" : '' : typeof quality === "string" && quality.includes("q_auto") ? quality : "q_auto:".concat(quality);

  if (!urlParams || !urlParams.length) {
    urlParams = 'c_lfill';
    if (fluid && !fluid.height) urlParams = 'c_scale';
  }

  if (!blurUrlParams || !blurUrlParams.length) {
    blurUrlParams = "c_scale,w_".concat(blurSize);
  } else {
    blurUrlParams = "".concat(blurUrlParams, ",w_").concat(blurSize);
  }

  if (useUrlParamsToBlur) {
    blurUrlParams = "".concat(urlParams, ",w_").concat(blurSize);
  }

  var toUrl = [imgFormat, quality, urlParams].filter(function (e) {
    return e && e.length;
  });
  var toBlurUrl = [imgFormat, quality, blurUrlParams].filter(function (e) {
    return e && e.length;
  });
  return {
    urlParams: toUrl.join(','),
    blurUrlParams: toBlurUrl.join(',')
  };
};

var detectWithAndHeight = function detectWithAndHeight(text) {
  var widthReg = new RegExp(/w_\d+/);
  var heightReg = new RegExp(/h_\d+/);
  return !!(text.match(widthReg) && text.match(heightReg));
}; // Cache if we've seen an image before so we don't both with
// lazy-loading & fading in on subsequent mounts.


var inImageCache = function inImageCache(props) {
  var image = props.fluid || props.fixed;

  var _makeUrlParams = makeUrlParams(props),
      urlParams = _makeUrlParams.urlParams;

  urlParams = props.fluid ? "".concat(urlParams, ",w_").concat(image.maxWidth).concat(image.height ? ",h_".concat(image.height) : '') : "".concat(urlParams, ",w_").concat(image.width, ",h_").concat(image.height); // Find src

  var src = "https://res.cloudinary.com/".concat(props.cloudName, "/image/upload/").concat(urlParams, "/").concat(props.version, "/").concat(props.imageName);

  try {
    var cache = JSON.parse(window.sessionStorage.getItem('seen_images')) || {};

    if (cache[src]) {
      return true;
    } else {
      cache[src] = true;
      window.sessionStorage.setItem('seen_images', JSON.stringify(cache));
      return false;
    }
  } catch (e) {
    return false;
  }
};

var io;
var listeners = [];

function getIO(IOParams) {
  if (typeof io === "undefined" && typeof window !== "undefined" && window.IntersectionObserver) {
    io = new window.IntersectionObserver(function (entries) {
      entries.forEach(function (entry) {
        listeners.forEach(function (l) {
          if (l[0] === entry.target) {
            // Edge doesn't currently support isIntersecting, so also test for an intersectionRatio > 0
            if (entry.isIntersecting || entry.intersectionRatio > 0) {
              io.unobserve(l[0]);
              l[1]();
            }
          }
        });
      });
    }, _objectSpread({}, IOParams));
  }

  return io;
}

var listenToIntersections = function listenToIntersections(el, IOParams, cb) {
  getIO(IOParams).observe(el);
  listeners.push([el, cb]);
};

var noscriptImg = function noscriptImg(props) {
  // Check if prop exists before adding each attribute to the string output below to prevent
  // HTML validation issues caused by empty values like width="" and height=""
  var src = props.src ? "src=\"".concat(props.src, "\" ") : "src=\"\" "; // required attribute

  var title = props.title ? "title=\"".concat(props.title, "\" ") : "";
  var alt = "alt=\"".concat(props.alt, "\""); // required attribute

  var width = props.width ? "width=\"".concat(props.width, "\" ") : "";
  var height = props.height ? "height=\"".concat(props.height, "\" ") : "";
  var opacity = props.opacity ? props.opacity : "1";
  var transitionDelay = props.transitionDelay ? props.transitionDelay : "0.5s";
  return "<img ".concat(width).concat(height).concat(src).concat(alt).concat(title, "style=\"position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:").concat(transitionDelay, ";opacity:").concat(opacity, ";width:100%;height:100%;object-fit:cover;object-position:center\"/>");
};

var Img = _react["default"].forwardRef(function (props, ref) {
  var style = props.style,
      onLoad = props.onLoad,
      onError = props.onError,
      otherProps = _objectWithoutProperties(props, ["style", "onLoad", "onError"]);

  return _react["default"].createElement("img", _extends({}, otherProps, {
    onLoad: onLoad,
    onError: onError,
    ref: ref,
    style: _objectSpread({
      position: "absolute",
      top: 0,
      left: 0,
      width: "100%",
      height: "100%",
      objectFit: "cover",
      objectPosition: "center"
    }, style)
  }));
});

Img.propTypes = {
  style: _propTypes["default"].object,
  onError: _propTypes["default"].func,
  onLoad: _propTypes["default"].func
};

var Image =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Image, _React$Component);

  function Image(props) {
    var _this;

    _classCallCheck(this, Image);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Image).call(this, props)); // If this browser doesn't support the IntersectionObserver API
    // we default to start downloading the image right away.

    var isVisible = true;
    var imgLoaded = true;
    var IOSupported = false;
    var fadeIn = props.fadeIn; // If this image has already been loaded before then we can assume it's
    // already in the browser cache so it's cheap to just show directly.

    var seenBefore = inImageCache(props);

    if (!seenBefore && typeof window !== "undefined" && window.IntersectionObserver) {
      isVisible = false;
      imgLoaded = false;
      IOSupported = true;
    } // Always don't render image while server rendering


    if (typeof window === "undefined") {
      isVisible = false;
      imgLoaded = false;
    }

    var hasNoScript = _this.props.fadeIn;
    _this.state = {
      isVisible: isVisible,
      imgLoaded: imgLoaded,
      IOSupported: IOSupported,
      fadeIn: fadeIn,
      hasNoScript: hasNoScript,
      seenBefore: seenBefore
    };
    _this.imageRef = _react["default"].createRef();
    _this.handleImageLoaded = _this.handleImageLoaded.bind(_assertThisInitialized(_this));
    _this.handleRef = _this.handleRef.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(Image, [{
    key: "handleRef",
    value: function handleRef(ref) {
      var _this2 = this;

      if (this.state.IOSupported && ref) {
        listenToIntersections(ref, this.props.IOParams, function () {
          _this2.setState({
            isVisible: true
          });
        });
      }
    }
  }, {
    key: "handleImageLoaded",
    value: function handleImageLoaded() {
      this.setState({
        imgLoaded: true
      });

      if (this.state.seenBefore) {
        this.setState({
          fadeIn: false
        });
      }

      this.props.onLoad && this.props.onLoad();
    }
  }, {
    key: "createBrakePointsFixed",
    value: function createBrakePointsFixed(urlCore) {
      var results = [];
      var image = this.props.fixed;

      for (var i = 1; i < 3; i++) {
        var params = "".concat(urlCore, ",w_").concat(image.width * i, ",h_").concat(image.height * i);
        results.push("https://res.cloudinary.com/".concat(this.props.cloudName, "/image/upload/").concat(params, "/").concat(this.props.version, "/").concat(this.props.imageName, " ").concat(i, "x"));
      }

      return results.join(',');
    }
  }, {
    key: "createBrakePointsFluid",
    value: function createBrakePointsFluid(urlCore) {
      var image = this.props.fluid;
      var step = image.step || 150;
      var size = 150;
      var results = [];

      while (size < image.maxWidth) {
        var params = "".concat(urlCore, ",w_").concat(size).concat(image.height ? ",h_".concat(Math.ceil(size * this.getAspectRatio(image))) : '');
        results.push("https://res.cloudinary.com/".concat(this.props.cloudName, "/image/upload/").concat(params, "/").concat(this.props.version, "/").concat(this.props.imageName, " ").concat(size, "w"));
        size = size + step;
      }

      results.push("https://res.cloudinary.com/".concat(this.props.cloudName, "/image/upload/").concat(urlCore, ",w_").concat(image.maxWidth).concat(image.height ? ",h_".concat(image.height) : '', "/").concat(this.props.version, "/").concat(this.props.imageName, " ").concat(image.maxWidth, "w"));
      return results.join(',');
    }
  }, {
    key: "getAspectRatio",
    value: function getAspectRatio(image) {
      return image.height > image.maxWidth ? image.height / image.maxWidth : image.maxWidth / image.height;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          title = _this$props.title,
          alt = _this$props.alt,
          cloudName = _this$props.cloudName,
          imageName = _this$props.imageName,
          _this$props$style = _this$props.style,
          style = _this$props$style === void 0 ? {} : _this$props$style,
          _this$props$imgStyle = _this$props.imgStyle,
          imgStyle = _this$props$imgStyle === void 0 ? {} : _this$props$imgStyle,
          _this$props$placehold = _this$props.placeholderStyle,
          placeholderStyle = _this$props$placehold === void 0 ? {} : _this$props$placehold,
          fluid = _this$props.fluid,
          fixed = _this$props.fixed,
          backgroundColor = _this$props.backgroundColor;

      var _makeUrlParams2 = makeUrlParams(this.props),
          urlParams = _makeUrlParams2.urlParams,
          blurUrlParams = _makeUrlParams2.blurUrlParams;

      var bgColor = typeof backgroundColor === "boolean" ? "lightgray" : backgroundColor;

      var imagePlaceholderStyle = _objectSpread({
        opacity: this.state.imgLoaded ? 0 : 1,
        transition: "opacity 0.5s",
        transitionDelay: this.state.imgLoaded ? "0.5s" : "0.25s"
      }, imgStyle, placeholderStyle);

      var imageStyle = _objectSpread({
        position: 'relative',
        opacity: this.state.imgLoaded || this.state.fadeIn === false ? 1 : 0,
        transition: this.state.fadeIn === true ? "opacity 0.5s" : "none"
      }, imgStyle);

      var placeholderImageProps = {
        title: title,
        alt: !this.state.isVisible ? alt : "",
        style: imagePlaceholderStyle
      };
      var image;
      var divStyle;
      var bgPlaceholderStyles;
      var srcSet;

      if (fluid) {
        image = fluid;
        divStyle = _objectSpread({
          position: "relative",
          overflow: "hidden",
          width: '100%',
          height: '100%'
        }, style);
        bgPlaceholderStyles = {
          backgroundColor: bgColor,
          position: "absolute",
          top: 0,
          bottom: 0,
          opacity: !this.state.imgLoaded ? 1 : 0,
          transitionDelay: "0.35s",
          right: 0,
          left: 0
        };
        srcSet = !detectWithAndHeight(urlParams) && this.createBrakePointsFluid(urlParams);
        urlParams = detectWithAndHeight(urlParams) ? urlParams : "".concat(urlParams, ",w_").concat(image.maxWidth).concat(image.height ? ",h_".concat(image.height) : '');
      }

      if (fixed) {
        image = fixed;
        divStyle = _objectSpread({
          position: "relative",
          overflow: "hidden",
          display: "inline-block",
          width: image.width,
          height: image.height
        }, style);
        bgPlaceholderStyles = {
          backgroundColor: bgColor,
          width: image.width,
          height: image.height,
          opacity: !this.state.imgLoaded ? 1 : 0,
          transitionDelay: "0.25s"
        };
        srcSet = !detectWithAndHeight(urlParams) && this.createBrakePointsFixed(urlParams);
        urlParams = detectWithAndHeight(urlParams) ? urlParams : "".concat(urlParams, ",w_").concat(image.width, ",h_").concat(image.height);
      }

      if (style.display === "inherit") {
        delete divStyle.display;
      }

      if (fluid || fixed) {
        return _react["default"].createElement("div", {
          style: divStyle,
          ref: this.handleRef
        }, !bgColor && _react["default"].createElement(Img, _extends({
          src: "https://res.cloudinary.com/".concat(cloudName, "/image/upload/").concat(blurUrlParams, "/").concat(this.props.version, "/").concat(imageName)
        }, placeholderImageProps)), bgColor && _react["default"].createElement("div", {
          title: title,
          style: bgPlaceholderStyles
        }), this.state.isVisible && _react["default"].createElement(Img, {
          alt: alt,
          title: title,
          src: "https://res.cloudinary.com/".concat(cloudName, "/image/upload/").concat(urlParams, "/").concat(this.props.version, "/").concat(imageName),
          srcSet: srcSet || '',
          style: imageStyle,
          ref: this.imageRef,
          onLoad: this.handleImageLoaded,
          onError: this.props.onError
        }), this.state.hasNoScript && _react["default"].createElement("noscript", {
          dangerouslySetInnerHTML: {
            __html: noscriptImg(_objectSpread({
              alt: alt,
              title: title
            }, image))
          }
        }));
      }

      return null;
    }
  }]);

  return Image;
}(_react["default"].Component);

Image.defaultProps = {
  cloudName: process.env.CLOUD_NAME || process.env.REACT_APP_CLOUD_NAME,
  fadeIn: true,
  alt: "",
  version: "",
  imgFormat: true,
  quality: true,
  blurSize: 20,
  useUrlParamsToBlur: false,
  IOParams: {
    rootMargin: '200px'
  }
};

var fixedObject = _propTypes["default"].shape({
  width: _propTypes["default"].number.isRequired,
  height: _propTypes["default"].number.isRequired
});

var fluidObject = _propTypes["default"].shape({
  maxWidth: _propTypes["default"].number.isRequired,
  height: _propTypes["default"].number,
  step: _propTypes["default"].number
});

Image.propTypes = {
  fixed: fixedObject,
  fluid: fluidObject,
  urlParams: _propTypes["default"].string,
  fadeIn: _propTypes["default"].bool,
  title: _propTypes["default"].string,
  alt: _propTypes["default"].string,
  cloudName: _propTypes["default"].string,
  imageName: _propTypes["default"].string.isRequired,
  style: _propTypes["default"].object,
  imgStyle: _propTypes["default"].object,
  placeholderStyle: _propTypes["default"].object,
  backgroundColor: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].bool]),
  onLoad: _propTypes["default"].func,
  onError: _propTypes["default"].func,
  imgFormat: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].bool]),
  quality: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].bool]),
  version: _propTypes["default"].string,
  blurSize: _propTypes["default"].number,
  blurUrlParams: _propTypes["default"].string,
  useUrlParamsToBlur: _propTypes["default"].bool,
  IOParams: _propTypes["default"].object
};
var _default = Image;
exports["default"] = _default;
